pkg/k8s/watchers/cilium_cidr_group.go

```diff
@@ -0,0 +1,281 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	cilium_v2_alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/k8s/watchers/resources"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+func (k *K8sWatcher) onUpsertCIDRGroup(
+	cidrGroup *cilium_v2_alpha1.CiliumCIDRGroup,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+	apiGroup, metricLabel string,
+) error {
+	var (
+		errs []error
+
+		equal  bool
+		action string
+	)
+
+	// wrap k.K8sEventReceived call into a naked func() to capture equal in the closure
+	defer func() {
+		k.K8sEventReceived(apiGroup, metricLabel, action, true, equal)
+	}()
+
+	oldCidrGroup, ok := cidrGroupCache[cidrGroup.Name]
+	if ok && oldCidrGroup.Spec.DeepEqual(&cidrGroup.Spec) {
+		equal = true
+		return nil
+	} else if ok {
+		action = resources.MetricUpdate
+	} else {
+		action = resources.MetricCreate
+	}
+
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	for key, cnp := range cnpCache {
+		if !hasCIDRGroupRef(cnp, cidrGroup.Name) || cnp.RequiresDerivative() {
```

Comment:
nit: I'm not sure I understand the logical impact of `RequiresDerivative()`, which seems to boil down to "Does this policy have a ToGroups statement?". That check seems to only apply to egress policies but this is an ingress policy feature, and just generally the notion of whether there is some such statement _somewhere_ in the CNP governing whether _this_ feature should work doesn't make sense to me. Is there some background to this that I'm missing?

```suggestion
		if !hasCIDRGroupRef(cnp, cidrGroup.Name) {
```

Reply:
No, I agree that it has no impact on the Ingress part of the policy. I'm gonna remove the check for `RequiresDerivative()`.